plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/TransferSemaphoresHolder.java

```diff
@@ -0,0 +1,143 @@
+            super(permits);
+            this.type = type;
+        }
+
+        public String getType() {
+            return type;
+        }
+    }
+
+    /**
+     * For multiple part requests of a single file, request context object will be set with the decision if low
+     * priority permits can also be utilized in high priority transfers of parts of the file. If high priority get fully
+     * consumed then low priority permits will be acquired for transfer.
+     *
+     * If a low priority transfer request comes in and a high priority transfer is in progress then till current
+     * high priority transfer finishes, low priority transfer may have to compete. This is an acceptable side effect
+     * because low priority transfers are generally heavy and it is ok to have slow progress in the beginning.
+     *
+     */
+    public static class RequestContext {
+
+        private final boolean lowPriorityPermitsConsumable;
+
+        private RequestContext(boolean lowPriorityPermitsConsumable) {
+            this.lowPriorityPermitsConsumable = lowPriorityPermitsConsumable;
+        }
+
+    }
+
+    public RequestContext createRequestContext() {
+        return new RequestContext(this.lowPrioritySemaphore.availablePermits() == lowPriorityPermits);
+    }
+
+    /**
+     * Acquire permit based on the availability and based on the transfer priority.
+     * A high priority event can acquire a low priority semaphore if all low permits are available.
+     * A low priority event can acquire a high priority semaphore if at least 40% of high permits are available. We
+     * reserve this bandwidth to ensure that high priority events never wait for permits in case of ongoing low priority
+     * transfers.
+     */
+    public TypeSemaphore acquirePermit(WritePriority writePriority, RequestContext requestContext) throws InterruptedException {
+        // Try acquiring low priority permit or high priority permit immediately if available.
+        // Otherwise, we wait for low priority permit.
+        if (Objects.requireNonNull(writePriority) == WritePriority.LOW) {
+            if (lowPrioritySemaphore.tryAcquire()) {
+                return lowPrioritySemaphore;
+            } else if (highPrioritySemaphore.availablePermits() > 0.4 * highPriorityPermits && highPrioritySemaphore.tryAcquire()) {
+                return highPrioritySemaphore;
+            } else if (lowPrioritySemaphore.tryAcquire(acquireWaitDuration, acquireWaitDurationUnit)) {
+                return lowPrioritySemaphore;
+            }
+            return null;
+        }
+
+        // Try acquiring high priority permit or low priority permit immediately if available.
+        // Otherwise, we wait for high priority permit.
+        if (highPrioritySemaphore.tryAcquire()) {
+            return highPrioritySemaphore;
+        } else if (requestContext.lowPriorityPermitsConsumable && lowPrioritySemaphore.tryAcquire()) {
+            return lowPrioritySemaphore;
```

Comment:
can we live without this ? or check if 40% of low priority permits  are available, to just make things simple ?

Reply:
Low priority permits are anyways just 30% of total permits. So 40% of 30% would be too less and won't help much if high priority permits are busy. Ability to acquire the entire low permits can only help. 
By setting up a request context we are in a way committing for a single large file for high priority upload that we will also use low priority permits in this case. No other high priority upload can acquire low permits after this. If we want to allow other high priority uploads also then we will have to keep track of on-going and waiting uploads and increment/decrement counts in synchronized block. So, our approach actually is simple and safe to provide ability to utilize low priority permits.

Reply:
But in this case, only 1 part will be getting permit. Rest of the parts would still wait for `highPrioritySemaphore` only , which is defeating the purpose . Is my understanding correct ?

Reply:
No, not 1 part. All parts of the file we take the decision before creating part streams. Check [this](https://github.com/opensearch-project/OpenSearch/pull/12159/files#diff-865f98e22f27e46f34144f074fbe290fbcd93c3ae52d13ddd0d63661123cccc5R81)